index.bs

```diff
@@ -12346,12 +12364,60 @@ It is the responsibility of specifications using Web IDL to state
 which global environment (or, by proxy, which global object) each platform
 object is associated with.
 
+<div algorithm>
+  To create a <dfn export>new</dfn> object implementing the interface |interface|, with a Realm
+  |realm|, perform the following steps:
+
+    1.  Return the result of creating a [=new-internal|new=] object implementing |interface|, with
+        |realm| and no target.
+</div>
+
+<div algorithm>
+  To create a <dfn lt=new-internal>new</dfn> object implementing the interface |interface|, with
+  a Realm |realm| and optionally a JavaScript function value |newTarget|, perform the following steps:
+
+    1.  Assert: |interface| is [=exposed=] in |realm|.
+    1.  Let prototype be the [=interface prototype object=] for |interface| in
+        |realm|.
+    1.  If |newTarget| is provided, then:
+        1.  Let |newTargetPrototype| be [=?=] [$Get$](|newTarget|, "prototype").
+        1.  If [$Type$](|newTargetPrototype|) is Object, then set |prototype| to
+            |newTargetPrototype|.
+    1.  Let |instance| be a newly created [=ECMAScript/object=] in |realm|.
+    1.  Set |instance|.\[[Prototype]] to |prototype|.
+    1.  Set |instance|'s essential internal methods to the definitions specified in
+        [=ECMA-262 Ordinary object internal methods and internal slots=].
+    1.  Let |interfaces| be the [=inclusive inherited interfaces=] of |interface|.
+    1.  [=list/iterate|For every=] [=interface=] |ancestor interface| in |interfaces|:
+        1.  [=Define the unforgeable regular operations=] of |ancestor interface| on |instance|,
+            given |realm|.
+        1.  [=Define the unforgeable regular attributes=] of |ancestor interface| on |instance|,
+            given |realm|.
+    1.  If |interface| is declared with the [{{Global}}] [=extended attribute=], then:
+        1.  [=Define the regular operations=] of |interface| on |instance|, given |realm|.
+        1.  [=Define the regular attributes=] of |interface| on |instance|, given |realm|.
+        1.  Set |instance|.\[[SetPrototypeOf]] as defined in [[#platform-object-setprototypeof]].
+    1.  Otherwise, if |interfaces| contains an [=interface=] which
+        [=support indexed properties|supports indexed properties=],
+        [=support named properties|named properties=], or both:
+        1.  Set |instance|.\[[GetOwnProperty]] as defined in
+            [[#legacy-platform-object-getownproperty]].
+        1.  Set |instance|.\[[Set]] as defined in
+            [[#legacy-platform-object-set]].
+        1.  Set |instance|.\[[DefineOwnProperty]] as defined in
+            [[#legacy-platform-object-defineownproperty]].
+        1.  Set |instance|.\[[Delete]] as defined in
+            [[#legacy-platform-object-delete]].
+        1.  Set |instance|.\[[PreventExtensions]] as defined in
+            [[#legacy-platform-object-preventextensions]].
+        1.  Set |instance|.\[[OwnPropertyKeys]] as defined in
+            [[#legacy-platform-object-ownpropertykeys]].
```

Comment:
This is slightly problematic, as I think we shouldn't access any internal methods (which define algorithms do) before they are fully initialized – at least ES doesn't. However, it seems certainly correct that the define algorithms should not be visible to these overridden internal methods.

Could we do something like
- Initialize these internal methods right after "Set *instance*'s essential internal methods"
- Use ValidateAndApplyPropertyDescriptor directly (or otherwise directly access the property table) instead of DefinePropertyOrThrow

I feel like this also matches implementations better, as I don't think they would directly call the internal methods during bootstrapping.

This also applies to the [[SetPrototypeOf]] assignment for [[Global]].

Reply:
This is already going wrong in the current spec text, whichever way you read it. I'm happy to fix the "define the attributes/operations" algorithms in a followup, but that's scope creep for this PR.